db/db_impl/db_impl_write.cc

```diff
@@ -161,75 +169,42 @@ Status DBImpl::MultiBatchWriteImpl(const WriteOptions& write_options,
-    auto version_set = versions_->GetColumnFamilySet();
-    WriteBatchInternal::AsyncInsertInto(
-        &writer, writer.sequence, version_set, &flush_scheduler_,
-        ignore_missing_faimly, this, &write_thread_.write_queue_);
-    // Because `LaunchParallelMemTableWriters` has add `write_group->size` to `running`,
-    // the value of `running` is always larger than one if the leader thread does not
-    // call `CompleteParallelMemTableWriter`.
-    while (writer.write_group->running.load(std::memory_order_acquire) > 1) {
-      // Write thread could exit and block itself if it is not a leader thread.
-      if (!write_thread_.write_queue_.RunFunc() && !is_leader_thread) {
-        break;
-      }
-    }
-    // We only allow leader_thread to finish this WriteGroup because there may
-    // be another task which is done by the thread that is not in this WriteGroup,
-    // and it would not notify the threads in this WriteGroup. So we must make someone in
-    // this WriteGroup to complete it and leader thread is easy to be decided.
-    if (is_leader_thread) {
-      if (!write_thread_.CompleteParallelMemTableWriter(&writer)) {
-        return Status::Aborted("Leader thread must complete at last and exit as memtable writer.");
-      }
-      MemTableInsertStatusCheck(writer.status);
-      versions_->SetLastSequence(writer.write_group->last_sequence);
-      write_thread_.ExitAsMemTableWriter(&writer, *writer.write_group);
-    } else {
-      write_thread_.CompleteParallelMemTableWriter(&writer);
+    if (writer.status.ok()) {
+      pending_memtable_writes_ += memtable_write_cnt;
     }
+    write_thread_.ExitAsBatchGroupLeader(wal_write_group, writer.status);
   }
 
   if (seq_used != nullptr) {
     *seq_used = writer.sequence;
   }
-  assert(writer.state == WriteThread::STATE_COMPLETED);
+  TEST_SYNC_POINT("DBImpl::WriteImpl:CommitAfterWriteWAL");
+
+  if (writer.request->commit_lsn != 0 && writer.status.ok()) {
+    TEST_SYNC_POINT("DBImpl::WriteImpl:BeforePipelineWriteMemtable");
+    size_t total_count = WriteBatchInternal::Count(my_batch);
+    InternalStats* stats = default_cf_internal_stats_;
+    stats->AddDBStats(InternalStats::kIntStatsNumKeysWritten, total_count);
+    RecordTick(stats_, NUMBER_KEYS_WRITTEN, total_count);
+
+    ColumnFamilyMemTablesImpl column_family_memtables(
+        versions_->GetColumnFamilySet());
+    writer.status = WriteBatchInternal::InsertInto(
+        &writer, writer.sequence, &column_family_memtables, &flush_scheduler_,
+        write_options.ignore_missing_column_families, 0 /*log_number*/, this,
+        true /*concurrent_memtable_writes*/);
+
+    WriteStatusCheck(writer.status);
+
+    if (!writer.FinalStatus().ok()) {
+      writer.status = writer.FinalStatus();
+    }
+    CommitSequence(writer.request);
+  } else if (writer.request->commit_lsn != 0) {
+    CommitSequence(writer.request);
+  } else {
+    writer.request->finished_write.store(true, std::memory_order_release);
```

Comment:
No need?

Reply:
Yes. But I think it is not a good idea to leave a request with `finished_write` false.